File: base/yield_demo.py
import asyncio;


def count():
    yield 1
    yield 2
    yield 3

gen = count()
print(next(gen))  # 输出: 1
print(next(gen))  # 输出: 2
print(next(gen))  # 输出: 3
# print(next(gen))  # error



def consumer():
    value = yield "Start"
    print(f"Received: {value}")
    yield "End"
    yield "End2"

gen = consumer()
val1 = next(gen)        # 启动生成器
print(f"val1:{val1}") # Start
val2 = gen.send("Data") # 输出: Received: Data, 执行yield 返回End
print(f"val2:{val2}")
print(next(gen)) # End2



gen2 = consumer()
val2 = next(gen2)        # 启动生成器
print(val2) # Start
# 输出: Received: None
print(next(gen2)) # End
print(next(gen2)) # End2




async def async_gen():
    yield "Start"
    await asyncio.sleep(3)
    yield "End"

async def run():
    async for value in async_gen():
        print(value)

asyncio.run(run())

File: base/yield_demo.js
function* count() {
  yield 1
  yield 2
  yield 3
}

const gen = count()
console.log(gen.next().value) // 输出: 1
console.log(gen.next().value) // 输出: 1
console.log(gen.next().value) // 输出: 1

function* consumer() {
  const value = yield "Start"
  console.log(`Received: ${value}`)
  yield "End"
}

{
  const gen = consumer()
  const val1 = gen.next() // 输出: { value: "Start", done: false }
  console.log(val1)
  const val2 = gen.next("Data") // 输出: Received: Data
  console.log(val2)
}
